content/posts/websocket-sample.md

```diff
@@ -0,0 +1,237 @@
+```
+
+- There you must have the `specs` directory. 
+We are going to use these specs and deploy our function.
+
+```
+fission spec apply
+
+DeployUID: 1df9464b-5f73-4623-8187-a2f431d5c828
+Resources:
+ * 2 Functions
+ * 2 Environments
+ * 2 Packages
+ * 2 Http Triggers
+ * 0 MessageQueue Triggers
+ * 0 Time Triggers
+ * 0 Kube Watchers
+ * 2 ArchiveUploadSpec
+Validation Successful
+Spec doesn't belong to Git Tree.
+2 environments created: nodejs, python
+2 packages created: broadcast-pkg, web-pkg
+2 functions created: broadcast, web
+2 HTTPTriggers created: broadcast, web
+```
+
+Now that we have deployed the function, let's test our chat application.
+Open multiple browser windows and go to `http://<router-ip:router-svc-port>/chat/app.html` to access the application.
+Send message from either of the window, it will be broadcasted to all others.
+
+### Manually generate specs
+
+- Clone the [fission/examples][6] repository and navigate to websocket sample code.
+
+```
+git clone https://github.com/fission/examples.git
+cd examples/samples/websocket
+```
+
+- Remove the existing specs directory
+```
+rmdir specs
+```
+
+- Create an initial declarative application specification.
+The command will recreate a `specs` directory and a fission deployment config spec file.
+
+```
+fission spec init
+```
+
+### Spec for broadcast function
+
+- Create fission [environment][7] for the broadcast function.
+
+```
+fission env create \
+ --name=nodejs \
+ --image=fission/node-env:latest \
+ --spec
+```
+
+- Create spec for the broadcast [function][8].
+
+```
+fission fn create \
+ --name=broadcast \
+ --env=nodejs \
+ --rpp=5 \
+ --deploy=broadcast.js \
+ --spec
+```
+
+- Create spec for [HTTP trigger][4].
+
+```
+fission httptrigger create \
+ --name=broadcast \
+ --url=/broadcast \
+ --function=broadcast \
+ --spec
+```
+
+### Spec for web function
+
+- Create fission [environment][7] for the web function.
+
+```
+fission env create \
+ --name=python \
+ --image=fission/python-env:latest \
+ --spec
+```
+
+- Create fission [package][14] for the web function.
+
+```
+fission package create \
+ --name=web-pkg \
+ --env=python \
+ --deploy="web/*" \
+ --spec
+```
+
+- Create spec for the web [function][8].
+
+```
+fission fn create \
+ --name=web \
+ --env=python \
+ --pkg=web-pkg \
+ --entrypoint=app.main \
+ --spec
+```
+
+- Create spec for [HTTP trigger][4].
+
+```
+fission httptrigger create \
+ --name=web \
+ --url='/chat/{html:[a-zA-Z0-9\.\/]+}' \
+ --function=web \
+ --spec
+```
+
+- Now, we can validate and apply the generated declarative application specifications.
+
+```
+fission spec apply
+```
+
+In order to test the application, open multiple browser windows and go to `http://<router-ip:router-svc-port>/chat/app.html` to access the application.
+Send message from either of the window, it will be broadcasted to all others.
+
+- You can cleanup and delete all Fission resources in the application specification by executing the command:
```

Comment:
One thing we should talk about is when the WebSocket is idle for a certain duration - the cleanup of pod happens and any parameter to tune that if available. It emphasises the fact that we are not running all containers all time - and only spinning them  up when there are requests.